src/napistu/ingestion/reactom_fi.py

```diff
@@ -0,0 +1,208 @@
+    annotated_interactions = interactions.merge(
+        formatted_annotations,
+        on=[REACTOME_FI.ANNOTATION, REACTOME_FI.DIRECTION],
+        how="left",
+    )
+
+    # flip reverse entries so all relationships are forward or undirected
+    formatted_interactions = (
+        pd.concat(
+            [
+                annotated_interactions.query("polarity == 'forward'"),
+                (
+                    annotated_interactions.query("polarity == 'reverse'").rename(
+                        columns={
+                            REACTOME_FI.GENE1: REACTOME_FI.GENE2,
+                            REACTOME_FI.GENE2: REACTOME_FI.GENE1,
+                        }
+                    )
+                ),
+            ]
+        )[[REACTOME_FI.GENE1, REACTOME_FI.GENE2, "sbo_term_name", "Score"]]
+        # looks like they were already unique edges
+        .sort_values("Score", ascending=False)
+        .groupby([REACTOME_FI.GENE1, REACTOME_FI.GENE2])
+        .first()
+    )
+
+    fi_edgelist = (
+        formatted_interactions.reset_index()
+        .rename(
+            columns={
+                REACTOME_FI.GENE1: "upstream_name",
+                REACTOME_FI.GENE2: "downstream_name",
+            }
+        )
+        .assign(r_Identifiers=Identifiers([]))
+    )
+
+    return fi_edgelist
+
+
+def _parse_reactome_fi_annotations(interactions: pd.DataFrame) -> pd.DataFrame:
+    """
+    Parse and annotate Reactome FI interaction types and directions using regex-based rules.
+
+    Parameters
+    ----------
+    interactions : pd.DataFrame
+        DataFrame containing Reactome FI interactions, with annotation and direction columns.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with annotation, direction, SBO term name, and polarity for each unique annotation/direction pair.
+
+    Raises
+    ------
+    ValueError
+        If an annotation/direction pair cannot be matched to a rule or if invalid directions are found.
+    """
+
+    distinct_annotations = (
+        interactions[[REACTOME_FI.ANNOTATION, REACTOME_FI.DIRECTION]]
+        .drop_duplicates()
+        .reset_index(drop=True)
+    )
+    invalid_directions = distinct_annotations.loc[
+        ~distinct_annotations[REACTOME_FI.DIRECTION].isin(VALID_REACTOME_FI_DIRECTIONS),
+        "Direction",
+    ]
+    if len(invalid_directions) > 0:
+        raise ValueError(f"Invalid directions: {invalid_directions}")
+
+    annotations = list()
+    for _, vals in distinct_annotations.iterrows():
+        annot, direction = vals
+
+        forward_match = utils.match_regex_dict(
+            annot, REACTOME_FI_RULES_FORWARD.NAME_RULES
+        )
+        if not forward_match:
+            if direction in REACTOME_FI_RULES_FORWARD.DIRECTION_RULES:
+                forward_match = REACTOME_FI_RULES_FORWARD.DIRECTION_RULES[direction]
+
+        reverse_match = utils.match_regex_dict(
```

Comment:
This does not need to be done in case the `forward_match` has already been successful, thus saving computation

If not refactored into a functional form (as suggested above), a append to annotations and use `continue` to skip to next loop iteration in case `forward_match` was sucessfull:
```
        if forward_match:
            annotations.append(
                {
                    REACTOME_FI.ANNOTATION: annot,
                    REACTOME_FI.DIRECTION: direction,
                    "sbo_term_name": forward_match,
                    "polarity": "forward",
                }
            )
            continue
   
```

Reply:
The weird thing about the Reactome FI annotations is that an annotation can be both a forward AND a reverse match. The arrow types are the best way of seeing this but for any pair of genes (A, B), they will not be defined also as (B, A).